core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+
+public class CachingCatalog implements Catalog {
+
+  public static Catalog wrap(Catalog catalog) {
+    return new CachingCatalog(catalog);
+  }
```

Comment:
Seems like user has to call the caching layer explicitly, is that what we want?

Reply:
Yes. We initially built the caching layer into our internal catalog without making it optional, but that was hard to work with. First, it was hard to debug when we had cache problems, like not invalidating the cache when a table is renamed. Now we have a catalog option (in DSv2 config) that turns caching on or off so we can see whether a problem is the cache. Second, this gives us a work-around for users hitting caching-related problems because they can turn it off.

Reply:
Do you consume that option through `initialize` in `CatalogPlugin`?

Reply:
Yes. We create the Iceberg `Catalog` in `initialize` and either wrap it with `CachingCatalog` or not:

Reply:
That seems nice. It would be great to do the same in our 3.0 integration.